### DISCUSSION TITLE
Why did CoqGym instrument/modify Coq?

### DISCUSSION THREAD

**Discussion by brando90** [princeton-vl/CoqGym#74] [Q&A] [answered]
Why did the CoqGym project instrument/modify Coq? 
I was reading a previous (excellent) discussion CoqDiscuss [https://coq.discourse.group/t/machine-learning-and-hammers-for-coq/303/34?u=brando90] and it says that it was needed to get local context and environment (but searched through the original coqgym paper and coqgym repo to confirm but couldn't). 

So I wanted to get an official/correct confirmation. Why CoqGym modify/instrument Coq? Was this really needed?

Thanks in advance for your time!

PS: for reference/completeness Proverbo9001 did not instrument/modify coq: https://github.com/HazardousPeach/coq_serapy to y best understanding.

ps2: related links/discussions:
- https://github.com/princeton-vl/CoqGym/discussions/70
- https://github.com/princeton-vl/CoqGym/discussions/71
- https://github.com/HazardousPeach/coq_serapy/issues/2
- https://github.com/UCSD-PL/proverbot9001/issues/26
- https://github.com/princeton-vl/CoqGym/discussions/74
- https://coq.discourse.group/t/machine-learning-and-hammers-for-coq/303/37
- https://github.com/UCSD-PL/proverbot9001/issues/27

**Comment by yangky11** [comment, marked as answer]
Do you mean the modifications we made in https://github.com/princeton-vl/CoqGym/blob/master/coq/toplevel/vernac.ml?

They were mainly for recording the path information (PWD, LOAD_PATH, ML_PATH) when a Coq project is being built so that you can restore them before executing the Coq commands in that project.

**Reply by brando90** [reply]
I am trying to understand why you instrumented Coq. That is mainly what I am trying to find out.

**Reply by yangky11** [reply]
If the paths are incorrect, executing Coq commands will throw errors. I am not aware of an alternative method for reconstructing exactly the same paths when compiling the Coq projects.

**Reply by brando90** [reply]
I am curious and puzzled. Why couldn't you do this soley with Serapi?

**Reply by yangky11** [reply]
Because when building a Coq project by running `make`, it uses the default Coq installation rather than SerAPI.

**Reply by brando90** [reply]
so what is SerAPI for then?

**Reply by yangky11** [reply]
For executing Coq scripts in Python (via serapi.py). This is after all Coq projects have been compiled.

**Reply by brando90** [reply]
Sorry if these are naive questions.

But why do you need to run `make` at all? Why can't you do everything with SerAPI?

**Reply by yangky11** [reply]
It usually wouldn't work if you just try to execute every Coq script directly using SerAPI. Running `make` ensures the Coq project is compiled in a way designed by its authors, which may include installing dependencies, compiling Coq files following a certain order, etc. 

It's analogous to runing `pip install XXX` to install a Python project instead of just using the Python interpreter to run every single Python file.

**Reply by brando90** [reply]
> It usually wouldn't work if you just try to execute every Coq script directly using SerAPI. Running `make` ensures the Coq project is compiled in a way designed by its authors, which may include installing dependencies, compiling Coq files following a certain order, etc.
> 
> It's analogous to runing `pip install XXX` to install a Python project instead of just using the Python interpreter to run every single Python file.

got it! Fantastic that was very useful.

I assume that this `make` (~ `pip install for coq`) is similar to evaluation of provers/models? We need to compile them before running the SerAPI tool? 

Do you run make using python or how to you make sure your model is ready to be evaluated (or to generate the data set)?

**Reply by yangky11** [reply]
I'm not sure if I understand the question correctly. `make` is run in step 5 of https://github.com/princeton-vl/CoqGym#12-building-coq-serapi-coqhammer-and-the-coq-projects. It is a prerequisite for extracting the dataset or evaluating the model.

**Reply by brando90** [reply]
Curious @yangky11 is there a disadvantage to what proverbot did to get the CoqGym code using the `--prelude` flag?

[Quoting Alex's response to give context](https://coq.discourse.group/t/machine-learning-and-hammers-for-coq/303/41?u=brando90):

>  What I did was simply have the user specify a root directory for the project (with a --prelude flag), and then parse the _CoqProject or Make file in that directory to get the load and ML paths. This is sufficient to run the coq scripts without error.

**Reply by yangky11** [reply]
The differences are mostly implementation details. They have to parse the makefiles whereas we just run them as black boxes.

**Reply by brando90** [reply]
@yangky11 I'm curious Kaiyu, how did you deal with the fact that the coq projects have different dependencies? Did you create a different opam switch for each project?

I'm also curious how that is related to the `PWD, LOAD_PATH, ML_PATH` instrumentation. My understanding is that you can just install projects given that you activated the right switch. So if that is so why do you need those paths in your coqgym code?

**Reply by yangky11** [reply]
Only a single OPAM switch is created [here](https://github.com/princeton-vl/CoqGym#12-building-coq-serapi-coqhammer-and-the-coq-projects).

The paths are used in https://github.com/princeton-vl/CoqGym/blob/cde1f3f2cd0fa208ea44085e26cae269eec2bc1b/utils.py#L160

You can also search `set_paths ` in this repo to see how it is used.
